Convert example notebooks to Markdown files

bkktimber opened the pull request:
Convert all notebooks in [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples) to Markdown files (https://github.com/blackjax-devs/blackjax/issues/128). Could you review this PR? @rlouf 

 A few important guidelines and requirements before we can merge your PR:

 - [x] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [ ] There are tests covering the changes;
 - [x] The doc is up-to-date;
 - [ ] If I add a sampler I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

Consider opening a **Draft PR** if your work is still in progress but you would like some feedback from other contributors.

rlouf commented on examples/Introduction.md:
````diff
@@ -0,0 +1,256 @@
+states.position["loc"].block_until_ready()
+```
+
+In this example the result is a dictionnary and each entry has shape `(num_samples, num_chains)`. Here's how to access the samples of the second chains for `loc`:
````
It looks like there are some lines missing here?

rlouf replied:
@bkktimber

rlouf commented on examples/Pathfinder.md:
```diff
@@ -0,0 +1,281 @@
+
+
+* L-BFGS algorithm struggles with float32s and log-likelihood functions; it's suggested to use double precision numbers. In order to do that in `jax` a configuration variable needs to be set up at initialization time (see [here](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision))
+
```
The linebreaks are not necessary

rlouf commented on examples/TemperedSMC.md:
````diff
@@ -0,0 +1,343 @@
+
+We plot the resulting tempered density for 5 different values of $\lambda_k$: from $\lambda_k =1$ which correponds to the original density to $\lambda_k=0$. The lower the value of $\lambda_k$ the easier it is to sampler from the posterior log-density.
+
+```python pycharm={"name": "#%%\n"}
````
same

rlouf commented on examples/TemperedSMC.md:
````diff
@@ -0,0 +1,343 @@
+ax.legend(list(lambdas))
+```
+
+```python pycharm={"name": "#%%\n"}
````
here too

rlouf commented on examples/TemperedSMC.md:
```diff
@@ -0,0 +1,343 @@
+    name: python3
+---
+
+<!-- #region pycharm={"name": "#%% md\n"} -->
```
many things added by PyCharm in this file, please remove all of them

rlouf commented on examples/use_with_pymc.md:
```diff
@@ -0,0 +1,136 @@
+    name: blackjax
+---
+
+<!-- #region id="397995ab" -->
```
Please clean the file as the previous one.

rlouf commented on examples/use_with_tfp.md:
```diff
@@ -0,0 +1,208 @@
+
+# Use BlackJAX with TFP
+
+<!-- #region -->
```
same

rlouf commented:
This looks great, well done. It looks like some text editors added some unnecessary comments/ids, could you remove them manually please? Next we can work on running the notebooks in CI to include them in the documentation :)

bkktimber commented:
Sure 😃

bkktimber commented:
I removed most of the metadata from the markdown files and kept only notebooks' metadata at the top of each file.

I used `jupytext.toml` file to set jupytext config on my local. Should I include the config file in PR too?

rlouf commented:
> I removed most of the metadata from the markdown files and kept only notebooks' metadata at the top of each file.

Looks much better!

> I used `jupytext.toml` file to set jupytext config on my local. Should I include the config file in PR too?

If it is important for people who clone the repo to have the same experience running / keeping the notebooks in sync with the md files I would say yes.

bkktimber commented:
I added markdown conversion instruction and updated documentation in [docs/examples.rst](https://github.com/blackjax-devs/blackjax/blob/main/docs/examples.rst)

rlouf commented:
Great, Now we need to have the CI run the notebooks and display them with images in the documentation so people can visualize their content browsing the doc.

bkktimber commented:
I follow this [link](https://docs.readthedocs.io/en/stable/guides/jupyter.html#other-formats) to add rendering configuration in `docs/conf.py`. Not sure is there another place I need to change?

rlouf commented:
I don't know. Does it work when you build the doc locally?

bkktimber commented:
Just to drop an update. Rendering `matplotlib`'s plots from markdown don't seem to work. I check how [Matplotlib](https://github.com/matplotlib/matplotlib/tree/main/doc) and [Scipy](https://github.com/scipy/scipy/tree/main/doc/source) deal with their docs. they use `.py` files with `matplotlib.sphinxext.plot_directive`.

I'll try to follow those repos. Hope I can finish this issue by this weekend.

rlouf commented:
Did you manage to build the docs without the images at least?

bkktimber commented:
yes, I did. the docs without images look fine.

rlouf commented:
Nice! I'll take a look around too for the figures

rlouf commented:
I am wondering if you can't just convert the `.md` files to `.ipynb` in CI and then use `nbsphinx` to execute and add to the documentation?

bkktimber commented:
Oh, that works too. I thought we are going to remove all `.ipynb` for good.

That means we keep examples in Markdown format on git and convert them into notebooks for documentation. is that correct?

rlouf commented:
> That means we keep examples in Markdown format on git and convert them into notebooks for documentation. is that correct?

That sounds a like a good strategy to me. Notebooks are not under version control, results are displayed in the docs, everyone happy!

rlouf commented:
Any progress on this @bkktimber ?

bkktimber commented:
@rlouf Sorry for taking this way too long. I managed to convert all notebooks into MyST-NB markdown.

I added several dependencies to `requirement-doc.txt` to run the example notebooks. It's already bloated. I leave really big dependencies , for example, `TensorFlow` and `PyMC` for further discussion.

Below are the list of examples that were not rendered group by the reasons:

**Need Dependencies**
`SGLD` - `tensorflow_datasets`
`use_with_pymc` - `PyMC`
`use_with_tfp`- `tensorflow`

**Error**
`use_with_pymc` - `NameError: name 'infos' is not defined`
`HierarchicalBNN` - `SyntaxError: invalid syntax (4001417937.py, line 3)`

Could you suggest what should I do with those files above? Thank you

rlouf commented:
No worries! Yes the dependency file will be bloated, but hopefully we can get rid of some in the future; for instance I am not sure that `use_with_XXX` type examples deserve a notebook; a script should be enough. We'll see later and you can add `tensorflow-dataset`, `tensorflow` and `pymc` as dependencies for now.

Any idea what the other errors refer to?

bkktimber commented:
I fixed the error in `HierarchicalBNN` already. I deleted the wrong code blocks when I cleaned up the markdown file.

I'll install the other dependencies and add all examples to the documentation page.

bkktimber commented:
@rlouf All the examples are converted to Markdown. I run `make build-doc` and most of them were executed with the command!

There are 2 examples that would like to mention:
- `example_with_tfp` : `tfp` is currently not officially support on M1. So I could not run the last code block on my local and I believe this example should be fine when running on other platforms.
- `example_with_pymc` : I got `AttributeError: 'Model' object has no attribute 'compute_initial_point'` from `Configuring the model for BlackJax` section in the notebook. I checked with [the one on the `main` branch](https://github.com/blackjax-devs/blackjax/blob/main/examples/use_with_pymc.ipynb) and I found it was not executed. I'm not sure where the error I got came from. So I will leave this example out for now.

Do you have any suggestions on the points I mention above?

rlouf commented:
That's awesome! For `PyMC` there might have been an API change? Anyway, fixing the notebooks is a bit out of scope for this PR, so you can focus on having the resulting notebooks displayed in the doc and then we can open issues for the failing notebooks. It's also great that we finally have them run in CI.

bkktimber commented:
Thank you @rlouf. I cleaned up my commit and make sure every notebook in the `examples` except `example_with_pymc` are able to build in CI. I will submit this PR for review.

rlouf commented on Makefile:
```diff
@@ -11,4 +11,6 @@ release:
 
 
 build-docs:
+	# pip install -r requirements-doc.txt
+	# conda install -c conda-forge pymc
```
Is this needed ? If so why is it commented ?

bkktimber replied:
Remove `conda install -c conda-forge pymc` since `use_with_pymc.py` is removed from documentation. also uncommented dependencies installation.

rlouf commented:
This looks good! Two things:
1. Is it possible to **not** have `use_with_pymc`, `use_with_numpyro` and `use_with_tfp` in the built documentation? This way we can remove a few large dependencies in `requirements-docs.py`
2. I don't see anything related to building the documentation in CI, is that normal?

bkktimber commented:
> 1. Is it possible to not have use_with_pymc, use_with_numpyro and use_with_tfp in the built documentation? This way we can remove a few large dependencies in requirements-docs.py

I removed them in the latest commit.

> I don't see anything related to building the documentation in CI, is that normal?

Assuming all examples are in Markdown format before committing to the repo, all markdown - jupyter conversions are done in [examples/conf.py](https://github.com/blackjax-devs/blackjax/pull/243/files#diff-85933aa74a2d66c3e4dcdf7a9ad8397f5a7971080d34ef1108296a7c6b69e7e3). I also write steps to prepare example files to build documentation in [examples/README](https://github.com/blackjax-devs/blackjax/pull/243/files#diff-49aaa2819e35a856818ecec8c9fa7e1c79ad028d3f44bd749736353cfb51bac9)

With that assumption, I use `make build-docs` to build the documentation. I find `workflows/build_docs.yml` also uses the same commands used in the Makefile. So my understanding is that if building documentation with `Makefile` works, building documentation in CI with `build_docs.yml` should work without adding extra steps. Please correct me if I'm wrong.

rlouf commented:
Looks good. Before merging could you rebase your commits? There should be one commit for converting to `.md` and another one to build them in CI I believe.

bkktimber commented:
Got it. I rebased and pushed.

rlouf commented:
All right this looks fantastic. @junpenglao wdyt?

rlouf commented on examples/README.md:
```diff
@@ -0,0 +1,45 @@
+# Examples
+
+When you add a new sampling fuctions, please add an Exmaple Notebook to describe how the function is used. Before push your function and its corresponding notebook to `Main` branch, please convert the notebook into text file.
```
```suggestion
When you add a new sampling function, please add an example here to describe how the function is used. Blackjax uses [Jupytext](https://github.com/mwouts/jupytext) to store Jupyter notebooks as Markdown files.
```

rlouf commented on examples/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## Converting Notebook Files to Text Files
+
+Please convert your `.ipynb` to `markdown` before commit your example to this repo. You can use the command below:
```
```suggestion
If you implemented your example in a notebook you can convert your `.ipynb` file to Markdown using the command below:
```

rlouf commented on examples/README.md:
````diff
@@ -0,0 +1,45 @@
+jupytext example/your_example_notebook.ipynb --to myst
+```
+
+The script will produce the `Markdown` version of your example notebook. Please make sure you add only the `Markdown` version to your commit.
````
```suggestion
The script will produce the `Markdown` version of your example notebook. Please make sure to only commit the  `Markdown` file.
```

rlouf commented on examples/README.md:
```diff
@@ -0,0 +1,45 @@
+
+3. Verify the generated documentation in `docs/builds`
+
+### Execution Times
```
Just saw that, very useful!

bkktimber commented:
Thank you 😄 Should I commit your suggestions or leave them to you when the PR is merged?

rlouf commented:
I will apply my suggestions and rebase once @junpenglao has reviewed the changes.

junpenglao commented on docs/examples.rst:
```diff
@@ -4,15 +4,13 @@ Examples
+   examples/GP_EllipticalSliceSampler.md
+   examples/SGMCMC.md
+   examples/change_of_variable_hmc.md
+   examples/Pathfinder.md
```
Below are missing:
```
   examples/use_with_numpyro.md
   examples/use_with_pymc3.md
   examples/use_with_tfp.md
```

bkktimber replied:
@rlouf and I decided to remove them from the documentation. please see detail [here](https://github.com/blackjax-devs/blackjax/pull/243#issuecomment-1218567986).

junpenglao replied:
I see, thanks.

junpenglao commented on examples/README.md:
```diff
@@ -0,0 +1,45 @@
+
+When you add a new sampling fuctions, please add an Exmaple Notebook to describe how the function is used. Before push your function and its corresponding notebook to `Main` branch, please convert the notebook into text file.
+
+## Converting Notebook Files to Text Files
```
Could we add a few sentences on how to edit existing examples?

bkktimber replied:
Sure. Can we do something like this, what do you think?

## Converting Notebook Files to Text Files

Please convert your `.ipynb` to `markdown` before commit your example to this repo. You can use the command below:

```python
jupytext example/your_example_notebook.ipynb --to myst
```

Once the example file is converted to a Markdown file, you have two options for editing:

1. Edit the Markdown version as it is a regular Markdown file.
2. Edit the Notebook version and convert it to a Markdown file with the command above. `Jupytext` can handle the change if the example has the same file name.

The script will produce the `Markdown` version of your example notebook. Please make sure you add only the `Markdown` version to your commit.

junpenglao replied:
LG overall - BTW is there a way to convert the .md file back to .ipynb?

bkktimber replied:
`jupytext examples/example_file.md --to notebook` should do it.

junpenglao replied:
Nice - please go ahead and add these into the README

bkktimber replied:
Got it. Added and pushed.

junpenglao reviewed (approved):
some minor nit, great job!

rlouf commented:
@bkktimber It looks like at least one notebook wasn't fully converted and the markdown version is shorter than the notebook version. Could you double-check that every notebook has been fully converted?

bkktimber commented:
@rlouf not sure what was missing. I double-checked with the document on `main` and could not find any missing. I uploaded documentation built with this commit [here](http://blackjax-pr-128.s3-website-ap-southeast-1.amazonaws.com/) for reference.

Could you suggest the missing part, please?

rlouf commented:
I was thinking about `Introduction.md` but it is identical to the original notebook. My bad @bkktimber! At least this PR will make it easy to go over the notebooks and correct the text. I'll rebase the commits if needed and merge, and then we'll just need to look at the docs to see if we can admire your work.

rlouf commented:
Merging. Great job!

bkktimber commented:
No worry @rlouf! Let's see how it's doing 😁

rlouf commented:
A notebook fails in CI, I need to investigate.

rlouf commented:
Just a quick note to say that I've worked with the documentation recently and it's fantastic, so thank you again!

bkktimber commented:
It's my pleasure. Glad to know my contribution is helpful.
